### ISSUE TITLE
[Feature Request] HTML Source: More ways to filter links?

### ISSUE THREAD

**Issue by Rhys-T** [ImranR98/Obtainium#921]
**Describe the feature**

I'm trying to set up Obtainium to install the [Kustom](https://kustom.rocks) apps (KLWP/KWGT/KLCK). Using the HTML source as described in #792 does let Obtainium find APKs for the apps, but I can't seem to control _which_ APK it picks - e.g. limiting it to the AOSP variant, or only installing stable versions and not betas[^aospbeta]. All the download link `href`s follow the same pattern of <code>https\://kustom.rocks/download/K<var>XYZ</var>/<var>123</var></code>, so there's no way for a regular expression to distinguish them.

I have a couple of ideas of ways to detect which links are which, but I'm not sure which one would be most useful for other Obtainium users, or most practical to implement:

- One way would be to use a custom XPath to get the links, instead of `document.querySelectorAll('a')`. For instance, to get the AOSP variant, I could use `//strong[text()="AOSP"]/ancestor-or-self::li//a`, or maybe just `//a[text()[contains(.,'aosp')]]`. (CSS selectors might be useful for some apps too - and would be simpler to write when they worked - but the Kustom download pages aren't structured in a way that would let me use selectors for this.)
  
  The HTML parser library Obtainium is using doesn't seem to have a `document.evaluate` function for working with XPaths, like the real DOM does, but it looks like there's at least [one Dart XPath library](https://pub.dev/packages/xpath_selector) that can work with that parser. I haven't used it at all though.

- The other option I can think of would be a way to filter APKs based on the filename _after_ redirection. That would handle AOSP vs. Google Play, and at least some of beta vs. stable[^weirdbeta]. But it would also mean that _all_ the download links would have to be requested, at least far enough to see where they were redirecting to.
  
  (One more bit of Kustom-specific weirdness: their download URLs only redirect to the APKs properly when I do a GET request. A HEAD request gets redirected somewhere else entirely.)

[^aospbeta]: There aren't any non-beta AOSP builds at the moment - the developer only just got the AOSP builds working again. But you get the idea.
[^weirdbeta]: As of this writing, the KLWP page lists the last couple of 3.74 beta builds without `beta` in their name for some reason - and the actual download filenames say `release` instead of `beta`. Not sure what's going on there, but if they're really betas, the filename is not enough to distinguish them.

**Describe alternatives you've considered (if applicable)**

Those are the options I've come up with so far. I'm not sure which one should be the 'request' and which one should be the 'alternative', though.

**Comment by ImranR98** [comment]
> A HEAD request gets redirected somewhere else entirely.

That's unfortunate - it would have been simpler to use this (could do it in a way that would be re-usable for #771).

**Comment by Rhys-T** [comment]
Yeah… I was trying to come up with approaches that wouldn't be specific to that site, but I ran across that issue, and figured I'd mention it so you wouldn't have to go through the same "it works for the _browser_, why doesn't it work for _me_?" stuff that I did.

If I do `curl -I https://kustom.rocks/download/KLWP/939`, I get redirected to `https://help.kustom.rocks/search?q=download/KLWP/939`. If I add `-XGET`, it sends me to the correct azureedge URL for the actual APK instead. My guess is that whatever web server they're using only has a route configured for that download URL for GET, not for HEAD, and it's falling back to a default 'file not found' route that redirects to a help search.

The filename on the final APK url is `kapp-aosp-klwp-prod-release.apk` - not enough to get a version number from, but enough to tell it's the AOSP version. (Still not sure why that one says `release` when it's listed in the beta section…)

I suppose one option would be to include a toggle switch for "Don't use HEAD requests". It looks like `curl -I -XGET` just does a full GET request like normal, then closes the connection as soon as the response body starts coming in. I don't know if the HTTP client library you're using gives you enough control to be able to do that. It might involve switching to the lower-level `send`/`StreamedResponse` APIs? I'm not sure if that would be worth it or not, though. I don't know if there are any other app sites that have this problem.

(The situation here is a little bit different from #771. Here, a given download URL always redirects to the same version of the app, and you still need the HTML to find new versions.)
